pgdog/src/backend/pool/dns_cache.rs

```diff
@@ -0,0 +1,388 @@
+use tokio_util::sync::CancellationToken;
+use tracing::{debug, error, info, warn};
+
+// -------------------------------------------------------------------------------------------------
+// ------ Constant(s) ------------------------------------------------------------------------------
+
+const DEFAULT_TTL: Duration = Duration::from_secs(30);
+
+// -------------------------------------------------------------------------------------------------
+// ------ Singleton(s) -----------------------------------------------------------------------------
+
+static DNS_CACHE: Lazy<Arc<DnsCache>> = Lazy::new(|| {
+    Arc::new(DnsCache {
+        cache: Arc::new(RwLock::new(HashMap::new())),
+        hostnames: Arc::new(RwLock::new(HashSet::new())),
+        ttl: Arc::new(RwLock::new(DEFAULT_TTL)),
+    })
+});
+
+// -------------------------------------------------------------------------------------------------
+// ------ DnsCache :: Public interface -------------------------------------------------------------
+
+#[derive(Clone, Debug)]
+struct DnsCacheEntry {
+    ip: IpAddr,
+    resolved_at: Instant,
+    ttl: Duration,
+}
+
+pub struct DnsCache {
+    cache: Arc<RwLock<HashMap<String, DnsCacheEntry>>>,
+    hostnames: Arc<RwLock<HashSet<String>>>,
+    ttl: Arc<RwLock<Duration>>,
+}
+
+impl DnsCache {
+    /// Retrieve the global DNS cache instance.
+    pub fn global() -> Arc<DnsCache> {
+        DNS_CACHE.clone()
+    }
+
+    /// Resolve hostname to socket address string.
+    pub async fn resolve(&self, hostname: &str) -> Result<IpAddr, std::io::Error> {
+        if let Some(ip) = self.get_cached_ip(hostname) {
+            return Ok(ip);
+        }
+
+        // Track hostname for future refreshes.
+        {
+            let mut hostnames = self.hostnames.write().unwrap();
+            if hostnames.insert(hostname.to_string()) {
+                info!("DNS Cache :: Added hostname: {}", hostname);
+            }
+        }
+
+        let ip = self.resolve_and_cache(hostname).await?;
+        Ok(ip)
+    }
+
+    /// Update the TTL for new cache entries.
+    pub fn set_ttl(&self, ttl: Duration) {
+        let mut current_ttl = self.ttl.write().unwrap();
+        *current_ttl = ttl;
+    }
+
+    /// Start the background refresh loop.
+    /// Returns a CancellationToken that can be used to stop the loop (needed for testing).
+    /// In production, you can ignore the return value with `_`.
+    pub fn start_refresh_loop(self: &Arc<Self>) -> tokio_util::sync::CancellationToken {
+        let cancel_token = CancellationToken::new();
+        let cancel_clone = cancel_token.clone();
+
+        let cache_ref = Arc::clone(self);
+        tokio::spawn(async move {
+            loop {
+                let interval = cache_ref.ttl();
+
+                select! {
+                    _ = sleep(interval) => {
+                        cache_ref.refresh_all_hostnames().await;
+                    }
+                    _ = cancel_clone.cancelled() => {
```

Comment:
This feels like one of those infinite loops that don't need to  be cancelled. It'll just stop the the pooler is restarted.